### PR TITLE
Switch over pilog to use better proof trait: `Proof2`

### LINKED ISSUE #397: Address ownership requirements in pilog
- [ ] Apply the solution in #135 to the `pilog` proof (see epic for details)

### PR DESCRIPTION
**Pull request by gatoWololo**
Closes #397

### CODE REVIEW

File: src/zkp/pilog.rs [hunk 1]
```diff
@@ -371,30 +381,23 @@ mod tests {
...
-        let proof = PiLogProof::prove(
-            &input,
-            &ProverSecret::new(x.clone(), rho),
-            &(),
-            &mut transcript,
-            rng,
-        )?;
-        let transcript = Transcript::new(b"PiLogProof Test");
-        Ok((proof, input, transcript))
+        let proof =
```
Comment (marsella) [inline src/zkp/pilog.rs:387]: suggestion: use the `transcript()` method here.

File: src/presign/participant.rs [hunk 2]
```diff
@@ -993,7 +993,7 @@ impl PresignParticipant {
             .local_storage
             .retrieve::<storage::RoundOnePublicBroadcast>(message.from())?;
         let public = RoundThreePublic::try_from(message)?;
-        public.verify(
+        public.clone().verify(
```
Comment (marsella) [inline src/presign/participant.rs:996]: note: This seemed weird to me and I think we should change this API at some point, but for now I [made a comment](https://github.com/boltlabs-inc/tss-ecdsa/issues/352#issuecomment-1577346861) on the most relevant issue.

**Review by marsella — CHANGES_REQUESTED** [review]
Thanks! A couple notes here, the only blocking change is making sure the docs on `ProverSecret` are up to date.

File: src/zkp/pilog.rs [hunk 3]
```diff
@@ -109,15 +111,17 @@ impl CommonInput {
 }
 
 /// The prover's secret knowledge.
-#[derive(ZeroizeOnDrop)]
-pub(crate) struct ProverSecret {
+///
+/// Copying/Cloning references is harmless and sometimes necessary. So we
+/// implement Clone and Copy for this type.
+pub(crate) struct ProverSecret<'a> {
```
[the remark this answers is not in the document]
Comment (gatoWololo) [reply src/zkp/pilog.rs:117]: Whoops. You are right, no need to do it if it's not needed!

File: src/zkp/pilog.rs [hunk 4]
```diff
@@ -126,10 +130,10 @@ impl Debug for ProverSecret {
     }
 }
 
-impl ProverSecret {
+impl<'a> ProverSecret<'a> {
     /// Collect prover secrets for proving [`PiLogProof`].
-    pub(crate) fn new(plaintext: BigNumber, nonce: Nonce) -> Self {
-        Self { plaintext, nonce }
+    pub(crate) fn new(plaintext: &'a BigNumber, nonce: &'a Nonce) -> ProverSecret<'a> {
```
[the remark this answers is not in the document]
Comment (gatoWololo) [reply src/zkp/pilog.rs:135]: Oh interesting, I'm actually not super familiar with the exact way lifetimes are needed in this case. I originally wrote it like `Self<'a>` but the compiler complained:
```rust
error[E0109]: lifetime arguments are not allowed on self type
   --> src/zkp/pilog.rs:132:75
    |
132 |     pub(crate) fn new(plaintext: &'a BigNumber, nonce: &'a Nonce) -> Self<'a> {
    |                                                                      ---- ^^ lifetime argument not allowed
    |                                                                      |
    |                                                                      not allowed on self type
    |
note: `Self` is of type `zkp::pilog::ProverSecret<'a>`
   --> src/zkp/pilog.rs:114:19
    |
114 | pub(crate) struct ProverSecret<'a> {
    |                   ^^^^^^^^^^^^ `Self` corresponds to this type
...
130 | impl<'a> ProverSecret<'a> {
    | ------------------------- `Self` is on type `ProverSecret` in this `impl`
help: the `Self` type doesn't accept type parameters, use the concrete type's name `ProverSecret` instead if you want to specify its type parameters
    |
132 |     pub(crate) fn new(plaintext: &'a BigNumber, nonce: &'a Nonce) -> ProverSecret<'a> {
    |                                                                      ~~~~~~~~~~~~
```
So I changed it to `ProverSecret<'a>` but you are right that it works with just `Self`. So the lifetime is inferred? :mask:

[the remark this answers is not in the document]
Comment (gatoWololo) [reply src/zkp/pilog.rs:135]: Anyways... I changed it!

File: src/zkp/pilog.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (marsella) [reply src/zkp/pilog.rs]: Thank you for your commitment to documentation anyway :pray:

**Review by marsella — APPROVED** [review]
Thanks, looks good!

**Pull request merged**